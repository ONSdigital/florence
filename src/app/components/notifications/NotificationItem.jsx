import React from "react";
import PropTypes from "prop-types";
import clsx from "clsx";

const NotificationItem = props => (
    <li
        key={props.id + props.type}
        className={clsx("notifications__item", { visible: props.isVisible }, `notifications__item--${props.type}`)}
        data-testid={props.id + props.type}
    >
        {props.message}
        {props.buttons.map((button, i) => {
            return (
                <button key={`button-${i}`} className="notifications__button" onClick={button.onClick}>
                    {button.text}
                    {button.icon}
                </button>
            );
        })}
    </li>
);

NotificationItem.propTypes = {
    type: PropTypes.oneOf(["neutral", "warning", "positive"]).isRequired,
    id: PropTypes.string.isRequired,
    message: PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired,
    isVisible: PropTypes.bool,
    buttons: PropTypes.arrayOf(
        PropTypes.shape({
            onClick: PropTypes.func.isRequired,
            text: PropTypes.string.isRequired,
            icon: PropTypes.element,
        })
    ),
};

export default NotificationItem;
