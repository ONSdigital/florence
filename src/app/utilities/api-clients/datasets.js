import http from "../http";
import url from "../url";

const publishingDatasetControllerURL = "/dataset-controller";

export default class datasets {
    static create = (datasetID, body) => {
        return http.post(`/dataset/datasets/${datasetID}`, body).then(response => {
            return response;
        });
    };

    static get = datasetID => {
        return http.get(`/dataset/datasets/${datasetID}`).then(response => {
            return response;
        });
    };

    static getAll() {
        return http.get(`/dataset/datasets`).then(response => {
            return response;
        });
    }

    static getAllList() {
        return http.get(`${publishingDatasetControllerURL}/datasets`).then(response => {
            return response;
        });
    }

    static getInstance(instanceID) {
        return http.get(`/dataset/instances/${instanceID}`).then(response => {
            return response;
        });
    }

    static getEdition(datasetID, editionID) {
        return http.get(`/dataset/datasets/${datasetID}/editions/${editionID}`).then(response => {
            return response;
        });
    }

    static getEditions(datasetID) {
        return http.get(`/dataset/datasets/${datasetID}/editions`).then(response => {
            return response;
        });
    }

    static getVersion(datasetID, edition, version) {
        return http.get(`/dataset/datasets/${datasetID}/editions/${edition}/versions/${version}`).then(response => {
            return response;
        });
    }

    static getVersions(datasetID, edition) {
        return http.get(`/dataset/datasets/${datasetID}/editions/${edition}/versions`).then(response => {
            return response;
        });
    }

    static getLatestVersionForEditions = async (datasetID, editionsList) => {
        return new Promise(async (resolve, reject) => {
            const versionFetches = editionsList.map(edition => {
                return http
                    .get(`/dataset/datasets/${datasetID}/editions/${edition.id}/versions/${edition.latestVersion}`)
                    .then(response => {
                        return response;
                    })
                    .catch(error => {
                        reject(error);
                        return;
                    });
            });

            const allLatestVersions = await Promise.all(versionFetches)
                .then(version => {
                    return version;
                })
                .catch(error => {
                    console.error(error);
                    reject(error);
                    return;
                });
            resolve(allLatestVersions);
        });
    };

    static getLatestVersion(datasetID) {
        const datasetURL = `/dataset/datasets/${datasetID}`;

        return new Promise(async (resolve, reject) => {
            const dataset = await http.get(datasetURL).catch(error => reject(error));

            if (!dataset || !dataset.current || !dataset.current.links.latest_version) {
                resolve();
                return;
            }

            const latestVersionPath = dataset.current.links.latest_version.href.replace(dataset.current.links.self.href, "");
            const latestVersionURL = datasetURL + latestVersionPath;

            const latestVersion = await http.get(latestVersionURL).catch(error => reject(error));
            resolve(latestVersion);
        });
    }

    static getLatestVersionURL = async datasetID => {
        const datasetURL = `/dataset/datasets/${datasetID}`;
        return http
            .get(datasetURL)
            .then(response => {
                const dataset = response.next || response.current || response;
                return new URL(dataset.links.latest_version.href).pathname;
            })
            .catch(error => error);
    };

    static getVersionDimensions(datasetID, edition, version) {
        return http.get(`/dataset/datasets/${datasetID}/editions/${edition}/versions/${version}/dimensions`).then(response => {
            return response;
        });
    }

    static updateInstanceDimensions(instanceID, dimensions) {
        const body = { dimensions };
        return http.put(`/instances/${instanceID}`, body);
    }

    static updateDimensionLabelAndDescription(instanceID, dimension, name, description) {
        const body = {
            label: name,
            description: description,
        };

        return http.put(`/instances/${instanceID}/dimensions/${dimension}`, body, true).then(response => {
            return response;
        });
    }

    static updateVersionMetadata(datasetID, edition, version, metadata) {
        if (typeof metadata !== "object") {
            return Promise.reject({ status: 400 });
        }
        return http.put(`/dataset/datasets/${datasetID}/editions/${edition}/versions/${version}`, metadata, true).then(response => {
            return response;
        });
    }

    static updateInstanceEdition(instanceID, edition) {
        const body = {
            edition: edition,
        };
        return http.put(`/dataset/instances/${instanceID}`, body, true).then(response => {
            return response;
        });
    }

    static confirmEditionAndCreateVersion(instanceID, edition, metadata) {
        let body = {
            edition: edition,
            state: "edition-confirmed",
        };
        if (metadata && typeof metadata === "object") {
            body = {
                ...body,
                ...metadata,
            };
        }
        return http.put(`/dataset/instances/${instanceID}`, body, true);
    }

    static updateDatasetMetadata(datasetID, metadata) {
        return http.put(`/dataset/datasets/${datasetID}`, metadata, true).then(response => {
            return response;
        });
    }

    static getAllInstances() {
        return http.get(`/dataset/instances`).then(response => {
            return response;
        });
    }

    static getCompletedInstancesForDataset(datasetID) {
        return http.get(`/dataset/instances?dataset=${datasetID}&state=completed`).then(response => {
            return response;
        });
    }

    static getCompletedInstances() {
        return http.get(`/dataset/instances?state=completed`).then(response => {
            return response;
        });
    }

    static getNewVersionsAndCompletedInstances() {
        return http.get(`/dataset/instances?state=completed,edition-confirmed,associated`).then(response => {
            return response;
        });
    }

    static getEditMetadata(datasetID, editionID, versionID) {
        return http.get(`${publishingDatasetControllerURL}/datasets/${datasetID}/editions/${editionID}/versions/${versionID}`).then(response => {
            return response;
        });
    }

    static getCantabularMetadata(datasetID, cantabularDatasetId, lang) {
        return http.get(`/cantabular-metadata/dataset/${datasetID}/cantabular/${cantabularDatasetId}/lang/${lang}`);
    }

    static putEditMetadata(datasetID, editionID, versionID, body) {
        return http
            .put(`${publishingDatasetControllerURL}/datasets/${datasetID}/editions/${editionID}/versions/${versionID}`, body, true)
            .then(response => {
                return response;
            });
    }

    static getVersionsList(datasetID, editionID) {
        return http.get(`${publishingDatasetControllerURL}/datasets/${datasetID}/editions/${editionID}/versions`).then(response => {
            return response;
        });
    }

    static getEditionsList(datasetID) {
        return http.get(`${publishingDatasetControllerURL}/datasets/${datasetID}/editions`).then(response => {
            return response;
        });
    }
}
