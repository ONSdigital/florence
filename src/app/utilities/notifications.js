import { addNotification, removeNotification, toggleNotificationVisibility } from '../config/actions';
import { store } from '../config/store';

export default class notifications {

    /**
     * Updates state with new notification
     * @param {object} configuration of notification , as follows: 
     * {
     *      id [String]
     *      type [oneOf(["neutral", "warning"])]
     *      message [String]
     *      autoDismiss [Number]
     *      isDismissable [Bool]
     *      buttons [Array of objects]
     * }
     */
    static add(notification) {
        const config = {
            type: notification.type || "neutral",
            message: notification.message || "",
            id: notification.id || Date.now(),
            buttons: notification.buttons || [],
            isVisible: false
        }

        if (notification.autoDismiss && notification.autoDismiss > 0) {
            const timer = window.setTimeout(() => {
                window.clearTimeout(timer);
                this.remove(config.id);
            }, notification.autoDismiss);
        }

        if (notification.isDismissable) {
            config.buttons.push({
                text: "Close",
                onClick: function() {
                    notifications.remove(config.id)
                }
            });
        }

        store.dispatch(addNotification(config));
        
        // Set a timeout so browser doesn't try to render component without the animation
        const animationTimer = window.setTimeout(() => {
            store.dispatch(toggleNotificationVisibility(config.id));
            window.clearTimeout(animationTimer);
        }, 10);
    }

    /**
     * Removes notification from state
     * @param {string} notificationID 
     */
    static remove(notificationID) {
        store.dispatch(toggleNotificationVisibility(notificationID));

        // Set a timeout so browser doesn't try to render component without the animation
        const animationTimer = window.setTimeout(() => {
            store.dispatch(removeNotification(notificationID));
            window.clearTimeout(animationTimer);
        }, 30);
    }

}