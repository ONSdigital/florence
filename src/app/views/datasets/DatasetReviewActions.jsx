import React, { Component } from 'react';
import PropTypes from 'prop-types';

const propTypes = {
    areDisabled: PropTypes.bool,
    includeSaveLabels: PropTypes.bool,
    reviewState: PropTypes.string,
    userEmail: PropTypes.string.isRequired,
    lastEditedBy: PropTypes.string,
    onSubmit: PropTypes.func,
    onApprove: PropTypes.func,
    notInCollectionYet: PropTypes.bool
};

class DatasetReviewActions extends Component {
    constructor(props) {
        super(props);
    }

    renderSubmit() {
        return (
            <button id="submit-for-review" type="button" onClick={this.props.onSubmit} disabled={this.props.areDisabled} className="btn btn--positive">
                {this.props.includeSaveLabels ? 
                    "Save and submit for review"
                :
                    "Submit for review"
                }
            </button>
        )
    }

    renderApprove() {
        return (
            <button id="mark-as-reviewed" type="button" onClick={this.props.onApprove} disabled={this.props.areDisabled} className="btn btn--positive">
                {this.props.includeSaveLabels ? 
                    "Save and submit for approval"
                :
                    "Submit for approval"
                }
            </button>
        )
    }

    render() {
        if (this.props.reviewState === "reviewed") {
            return <span></span>;
        }

        if (this.props.notInCollectionYet || this.props.reviewState === "inProgress") {
            return this.renderSubmit();
        }
        
        if (this.props.userEmail === this.props.lastEditedBy && this.props.reviewState === "complete") {
            return this.renderSubmit();
        }

        if (this.props.userEmail !== this.props.lastEditedBy && this.props.reviewState === "complete") {
            return this.renderApprove();
        }

        return <span></span>;
        
    }
}

DatasetReviewActions.propTypes = propTypes;

export default DatasetReviewActions;