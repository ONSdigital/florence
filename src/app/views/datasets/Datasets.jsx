import React, { Component } from 'react';
import PropTypes from 'prop-types';

const propTypes = {
    datasets: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.string.isRequired,
        alias: PropTypes.string.isRequired
    })),
    onNewVersionClick: PropTypes.func.isRequired
}

class Datasets extends Component {
    constructor(props) {
        super(props);
    }

    render() {
        return (
            <div>
                {this.props.datasets.map(dataset => {
                    return (
                        // <Link to={`${this.props.rootPath}/datasets/${dataset.id}`} key={dataset.id}>
                        //     {dataset.alias}
                        // </Link>
                        <div className="grid grid--justify-space-between" key={dataset.id}>
                            {dataset.alias}
                            <button 
                                className="btn btn--primary" 
                                data-recipe-id={dataset.id}
                                onClick={this.onNewVersionClick}
                            >
                                Upload new version
                            </button>
                        </div>
                    );
                })}
            </div>
        )
    }
}

Datasets.propTypes = propTypes;

export default Datasets;