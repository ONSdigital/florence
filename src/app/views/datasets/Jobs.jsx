import React, { Component } from 'react';
import { Link } from 'react-router';
import PropTypes from 'prop-types';

const propTypes = {
    datasets: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.string.isRequired,
        alias: PropTypes.string.isRequired
    })),
    jobs: PropTypes.arrayOf(PropTypes.shape({
        job_id: PropTypes.string.isRequired,
        recipe: PropTypes.string.isRequired
    })),
    rootPath: PropTypes.string.isRequired
}

class Jobs extends Component {
    constructor(props) {
        super(props);
    }

    render() {
        return (
            <div>
                {this.props.jobs.length > 0 ?
                    <ul className="list">
                        {this.props.jobs.map(job => {
                            const recipe = this.props.datasets.find(dataset => {
                                return dataset.id === job.recipe;
                            });
                            if (!recipe) {
                                console.warn(`Attempt to render job for an unrecognised recipe ID: "${job.recipe}"` );
                                return ""
                            }
                            return (
                                <li className="list__item list__item--separated grid grid--justify-space-between" key={job.job_id}>
                                    <Link to={`${this.props.rootPath}/datasets/${job.recipe}/jobs/${job.job_id}`}>
                                        {recipe.alias}
                                    </Link>
                                    <span>{job.state}</span>
                                </li>
                            )
                        })}
                    </ul>
                :
                    <p>No dataset uploads in progress</p>
                }
            </div>
        )
    }
}

Jobs.propTypes = propTypes;

export default Jobs;