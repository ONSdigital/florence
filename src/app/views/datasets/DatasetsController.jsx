import React, { Component } from 'react';
import { connect } from 'react-redux';
import { push } from 'react-router-redux';
import PropTypes from 'prop-types';

import { updateAllDatasets, updateAllJobs, addNewJob } from '../../config/actions';
import recipes from '../../utilities/api-clients/recipes';
import datasetImport from '../../utilities/api-clients/datasetImport';
import notifications from '../../utilities/notifications';

import DatasetItem from './DatasetItem';
import Jobs from './Jobs';

const propTypes = {
    dispatch: PropTypes.func.isRequired,
    datasets: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.string.isRequired,
        alias: PropTypes.string
    })),
    jobs: PropTypes.arrayOf(PropTypes.shape({
        job_id: PropTypes.string.isRequired,
        recipe: PropTypes.string.isRequired
    })),
    rootPath: PropTypes.string.isRequired
}

export class DatasetsController extends Component {
    constructor(props) {
        super(props);

        this.state = {
            isFetchingData: false,
            disabledDataset: ""
        };

        this.handleNewVersionClick = this.handleNewVersionClick.bind(this);
    }
    
    componentWillMount() {
        this.setState({isFetchingData: true});

        const getRecipes = recipes.getAll();
        const getJobs = datasetImport.getAll();

        Promise.all([getRecipes, getJobs]).then(response => {
            this.props.dispatch(updateAllDatasets(response[0].items));
            this.props.dispatch(updateAllJobs(response[1]));
            this.setState({isFetchingData: false});
        }).catch(error => {
            switch (error.status) {
                case(403):{
                    const notification = {
                        "type": "warning",
                        "message": "You do not permission to view all datasets.",
                        isDismissable: true
                    }
                    notifications.add(notification)
                    break;
                }
                case("RESPONSE_ERR"):{
                    const notification = {
                        "type": "warning",
                        "message": "An error's occurred whilst trying to get all datasets.",
                        isDismissable: true
                    }
                    notifications.add(notification)
                    break;
                }
                case("FETCH_ERR"): {
                    const notification = {
                        type: "warning",
                        message: "There's been a network error whilst trying to get all datasets. Please check you internet connection and try again in a few moments.",
                        isDismissable: true
                    }
                    notifications.add(notification);
                    break;
                }
                case("UNEXPECTED_ERR"): {
                    const notification = {
                        type: "warning",
                        message: "An unexpected error has occurred whilst trying to get all datasets.",
                        isDismissable: true
                    }
                    notifications.add(notification);
                    break
                }
                default: {
                    const notification = {
                        type: "warning",
                        message: "An unexpected error's occurred whilst trying to get all datasets.",
                        isDismissable: true
                    }
                    notifications.add(notification);
                    break;
                }
            }
            this.setState({isFetchingData: false});
        });
    }

    shouldComponentUpdate(nextProps, nextState) {
        
        // Don't render all the jobs and datasets yet because the jobs haven't been added to the state yet
        if (nextProps.datasets.length > 0 && !nextProps.jobs && !nextState.isFetchingData) {
            return false;
        }
        
        return true;
    }

    handleNewVersionClick(event) {
        const recipeID = event.target.getAttribute('data-recipe-id');
        this.setState({disabledDataset: recipeID});
        datasetImport.create(recipeID).then(response => {
            this.props.dispatch(addNewJob(response));
            this.props.dispatch(push(`${this.props.rootPath}/datasets/${recipeID}/jobs/${response.job_id}`));
        }).catch(error => {
            this.setState({disabledDataset: ""});
            switch (error.status) {
                case(403):{
                    const notification = {
                        "type": "warning",
                        "message": "You do not have permission to upload a new version of this dataset.",
                        isDismissable: true
                    }
                    notifications.add(notification)
                    break;
                }
                case(404):{
                    const notification = {
                        "type": "warning",
                        "message": "This dataset was not recognised.",
                        isDismissable: true
                    }
                    notifications.add(notification)
                    break;
                }
                case("RESPONSE_ERR"):{
                    const notification = {
                        "type": "warning",
                        "message": "An error's occurred whilst trying to create a new version of this dataset.",
                        isDismissable: true
                    }
                    notifications.add(notification)
                    break;
                }
                case("FETCH_ERR"): {
                    const notification = {
                        type: "warning",
                        message: "There's been a network error whilst trying to create a new version of this dataset. Please check you internet connection and try again in a few moments.",
                        isDismissable: true
                    }
                    notifications.add(notification);
                    break;
                }
                case("UNEXPECTED_ERR"): {
                    const notification = {
                        type: "warning",
                        message: "An unexpected error has occurred whilst trying to create a new version of this dataset.",
                        isDismissable: true
                    }
                    notifications.add(notification);
                    break
                }
                default: {
                    const notification = {
                        type: "warning",
                        message: "An unexpected error's occurred whilst trying to create a new version of this dataset.",
                        isDismissable: true
                    }
                    notifications.add(notification);
                    break;
                }
            }
            console.error("Error creating new import job: ", error);
        });
    }

    render() {
        return (
            <div className="grid grid--justify-center">
                <div className="grid__col-4">
                    <h1>All datasets</h1>
                    <h2 className="margin-bottom--1">In progress</h2>
                    <div className="margin-bottom--2">
                        {this.state.isFetchingData &&
                            <div className="grid--align-self-start"> 
                                <div className="loader loader--dark"></div>
                            </div> 
                        }
                        {!this.state.isFetchingData &&
                            <Jobs jobs={this.props.jobs} datasets={this.props.datasets} rootPath={this.props.rootPath} />
                        }
                    </div>
                    <h2 className="margin-bottom--1">Datasets available to you</h2>
                    {this.state.isFetchingData &&
                        <div className="grid--align-self-start"> 
                            <div className="loader loader--dark"></div>
                        </div>
                    }
                    {this.props.datasets.length > 0 &&
                        <ul className="list">
                            {this.props.datasets.map(dataset => {
                                return (
                                    <DatasetItem 
                                        key={dataset.id} 
                                        dataset={dataset} 
                                        onNewVersionClick={this.handleNewVersionClick} 
                                        isLoading={dataset.id === this.state.disabledDataset}
                                    />
                                )
                            })}
                        </ul>
                    }
                </div>
            </div>
        )
    }
}

DatasetsController.propTypes = propTypes;

function mapStateToProps(state) {
    return {
        datasets: state.state.datasets.all,
        jobs: state.state.datasets.jobs,
        rootPath: state.state.rootPath
    }
}

export default connect(mapStateToProps)(DatasetsController);