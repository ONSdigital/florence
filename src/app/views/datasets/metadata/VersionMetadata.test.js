import React from 'react';
import {VersionMetadata} from './VersionMetadata.jsx';
import notifications from '../../../utilities/notifications';
import datasets from '../../../utilities/api-clients/datasets';
import uuid from 'uuid/v4';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';

console.error = jest.fn();
console.warn = jest.fn();

jest.mock('uuid/v4', () => () => {
    return "12345";
});

jest.mock('../../../utilities/url', () => {
    return {
        resolve: function() {
            return "test"
        },
        parent: function() {}
    }
});

jest.mock('../../../utilities/notifications', () => {
    return {
        add: jest.fn(() => {
            //
        })
    }
});

jest.mock('../../../utilities/http', () => {
    return {
        resolve: function() {
            //
        }
    }
});

jest.mock('../../../utilities/log', () => {
    return {
        add: function() {
            //
        },
        eventTypes: {}
    }
});

jest.mock('../../../utilities/api-clients/recipes', () => (
    {
        getAll: jest.fn(() => {
            return Promise.resolve({
                items:[]
            })
        })
    }
));

jest.mock('../../../utilities/api-clients/datasets', () => (
    {
        get: jest.fn(() => {
            return Promise.resolve({
                next: {}
            });
        }),
        getVersion: jest.fn(() => {
            return Promise.resolve({
            })
        }),
        getInstance: jest.fn(() => {
            return Promise.resolve({
            })
        })
    }
));

const mockEvent = {
    preventDefault: function() {}
}

const exampleDataset = {
    current: {
        collection_id: 'fddffdfdaadf-e8ad17766a81bf589e76ef57d854945fdf0bfe546000228837aa70701506869c',
        id: '931a8a2a-0dc8-42b6-a884-7b6054ed3b68',
        license: 'Open Government License',
        links: {
            editions: {
                href: 'http://localhost:22000/datasets/931a8a2a-0dc8-42b6-a884-7b6054ed3b68/editions'
            },
            latest_version: {
                id: 'efcc4581-30b1-463b-b85b-2e2d85c4918b',
                href: 'http://localhost:22000/datasets/931a8a2a-0dc8-42b6-a884-7b6054ed3b68/editions/2016/versions/1'
            },
            self: {
                href: 'http://localhost:22000/datasets/931a8a2a-0dc8-42b6-a884-7b6054ed3b68'
            }
        },
        next_release: 'pudding',
        keywords: [
            'keyword1',
            'keyword2'
        ],
        publisher: {},
        contacts: [{
            email: "test@email.com",
            name: "foo bar",
            telephone: "01633 123456"
        }],
        release_frequency: "Monthly",
        state: 'published',
        title: 'CPI',
        uri: '/economy/inflationandpricesindices/datasets/consumerpriceindices'
    }
}

const exampleInstance = {
    "id" : "b5655290-0414-4f94-a773-fdc14652d868",
    "dimensions" : [ 
        {
            "description" : "sdsds",
            "href" : "http://localhost:22400/code-lists/64d384f1-ea3b-445c-8fb8-aa453f96e58a",
            "id" : "64d384f1-ea3b-445c-8fb8-aa453f96e58a",
            "name" : "time",
            "label" : ""
        }, 
        {
            "description" : "dsdsdsd",
            "href" : "http://localhost:22400/code-lists/65107A9F-7DA3-4B41-A410-6F6D9FBD68C3",
            "id" : "65107A9F-7DA3-4B41-A410-6F6D9FBD68C3",
            "name" : "geography",
            "label" : ""
        }, 
        {
            "description" : "dssdsds",
            "href" : "http://localhost:22400/code-lists/e44de4c4-d39e-4e2f-942b-3ca10584d078",
            "id" : "e44de4c4-d39e-4e2f-942b-3ca10584d078",
            "name" : "aggregate",
            "label" : ""
        }
    ],
    "state" : "associated",
    "edition" : "time-series",
    "alerts" : [ 
        {
            "date" : "21/10/2018",
            "description" : "This an alert",
            "type" : "correction"
        }
    ],
    "latest_changes" : [ 
        {
            "description" : "Big change!",
            "name" : "Test change",
            "type" : "summary of changes"
        }
    ],
    "version" : 1,
    "collection_id" : "acceptancetestcollection-e329dda6d14eb3a280eb8536230ebdc1ee48751bccc4df87ee12059bcceae0e8"
}


const defaultProps = {
    dispatch: function(){},
    dataset: {...exampleDataset.current},
    instance: {...exampleInstance},
    version: {...exampleInstance},
    recipes: [{
        output_instances: [ 
            {
                editions: ["time-series", "another-type"],
                dataset_id: "931a8a2a-0dc8-42b6-a884-7b6054ed3b68",
            }
        ]
    }],
    params: {
        datasetID: "931a8a2a-0dc8-42b6-a884-7b6054ed3b68",
        version: "1",
        edition: "time-series"
    },
    location: {
        query: {collection: "a-collection-id"}
    },
    router: {
        listenBefore: () => {}
    }
}

test("Dataset version metadata page matches stored snapshot", () => {
    const component = renderer.create(
      <VersionMetadata {...defaultProps} />
    );
    expect(component.toJSON()).toMatchSnapshot();
});

test("Dataset title updates after successful fetch from dataset API on mount", async () => {
    const component = shallow(
        <VersionMetadata {...defaultProps} />
    );

    datasets.get.mockImplementationOnce(() => (
        Promise.resolve({current: exampleDataset.current}
    )));

    await component.instance().componentWillMount();
    await component.update(); // update() appears to be async so we need to wait for it to finish before asserting
    expect(component.state("title")).toBe("CPI");
});

test("Correct modal type shows when user wants to add an alert", async () => {
    const component = shallow(
        <VersionMetadata {...defaultProps} />
    );
    expect(component.state("modalType")).toBe(undefined);
    await component.instance().handleAddRelatedClick("alerts");
    expect(component.state("modalType")).toBe("alerts");
});

test("Handler for removing a summary of changes updates state correctly", async () => {
    const component = shallow(
        <VersionMetadata {...defaultProps} />
    );

    await component.instance().componentWillMount();
    await component.update();
    expect(component.state("alerts").length).toEqual(1);
    expect(component.state("alerts")[0]).toMatchObject({
        key: "12345",
        date: exampleInstance.alerts[0].date,
        description: exampleInstance.alerts[0].description,
        type: exampleInstance.alerts[0].type
    });
    await component.instance().handleDeleteRelatedClick("alerts", "12345");
    await component.update();
    expect(component.state("alerts").length).toBe(0);
});

test("Latest changes are set in state correctly on mount", async () => {
    const component = shallow(
        <VersionMetadata {...defaultProps} />
    );

    await component.instance().componentWillMount();
    await component.update();
    expect(component.state("changes").length).toEqual(1);
    component.state("changes").forEach((change, index) => {
        expect(change).toMatchObject({
            key: "12345",
            name: exampleInstance.latest_changes[0].name,
            description: exampleInstance.latest_changes[0].description,
            type: exampleInstance.latest_changes[0].type
        });
    });
});

test("Available editions maps correctly to select element", async () => {
    const component = shallow(
        <VersionMetadata {...defaultProps} />
    );

    const validSelectContents = ["time-series", "another-type"];
    await component.instance().componentWillMount();
    await component.update();
    const createdSelectContents = component.instance().mapEditionsToSelectOptions();
    expect(createdSelectContents).toEqual(expect.arrayContaining(validSelectContents));
});

test("Handle select change event updates state correctly", async () => {
    const component = shallow(
        <VersionMetadata {...defaultProps} />
    );

    await component.instance().componentWillMount();
    await component.update();
    expect(component.state("edition")).toBe("time-series");
    component.instance().handleSelectChange({preventDefault: ()=>{}, target: {id: "edition", value: "another-type"}});
    expect(component.update().state("edition")).toBe("another-type");
});

test("Changing an input value updates the state to show a change has been made", async () => {
    const component = await shallow(
        <VersionMetadata {...defaultProps} />
    );

    await component.update();
    expect(component.state("hasChanges")).toBe(false);
    component.instance().handleSelectChange({preventDefault: ()=>{}, target: {id: "edition", value: "another-type"}});
    expect(component.state("edition")).toBe("another-type");
    expect(component.state("hasChanges")).toBe(true);
});

test("Warning modal shown when unsaved changes have been made", async () => {
    const component = await shallow(
        <VersionMetadata {...defaultProps} />
    );

    await component.update();
    expect(component.find(".modal__header h2").exists()).toBe(false);
    expect(component.state("showModal")).toBe(false);
    component.instance().handleSelectChange({preventDefault: ()=>{}, target: {id: "edition", value: "another-type"}});
    component.instance().handleBackButton();
    expect(component.state("showModal")).toBe(true);
    expect(component.state("modalType")).toBe(undefined);
    expect(component.find(".modal__header h2").exists()).toBe(true);
});

test("Input errors are added on submit and then removed on change of that input", () => {
    const component = shallow(
        <VersionMetadata {...defaultProps} />
    );
    const mockTitleEvent = {
        target: {
            value: "Some free text",
            name: "add-related-content-title"
        }
    }
    const mockDescEvent = {
        target: {
            value: "Some description",
            name: "add-related-content-desc"
        }
    }

    expect(component.state("titleInput")).toBe("");
    expect(component.state("descInput")).toBe("");
    expect(component.update().state("titleError")).toBe(undefined);
    expect(component.update().state("descError")).toBe(undefined);

    component.instance().handleRelatedContentSubmit({preventDefault: ()=>{}});
    expect(component.update().state("titleError")).not.toBe("");
    expect(component.update().state("descError")).not.toBe("");

    component.instance().handleInputChange(mockTitleEvent);
    expect(component.state("titleInput")).toBe(mockTitleEvent.target.value);
    expect(component.update().state("titleError")).toBe(null);

    component.instance().handleInputChange(mockDescEvent);
    expect(component.state("descInput")).toBe(mockDescEvent.target.value);
    expect(component.update().state("descError")).toBe(null);
});

test("Handler to edit an alert updates the state with new values", async () => {
    const component = await shallow(
        <VersionMetadata {...defaultProps} />
    );
    await component.update();
    
    expect(component.state("alerts").length).toBe(1);
    const initialAlerts = component.state("alerts");
    component.setState({titleInput: "Some new words"});
    component.instance().editRelatedLink("alerts", "12345");
    expect(component.state("alerts").length).toBe(1);
    const editedAlerts = component.update().state("alerts");
    expect(editedAlerts[0].date).toBe("Some new words");
});

test("Handle for editing an opens modal with existing values", async () => {
    const component = await shallow(
        <VersionMetadata {...defaultProps} />
    );
    await component.update();
    
    component.instance().handleEditRelatedClick("alerts", "12345");
    expect(component.state("showModal")).toBe(true);
    expect(component.state("modalType")).toBe("alerts");
    expect(component.state("descInput")).toBe(exampleInstance.alerts[0].description);
    expect(component.state("titleInput")).toBe(exampleInstance.alerts[0].date);
});

test("Alert items map to card element correctly", async () => {
    const component = await shallow(
        <VersionMetadata {...defaultProps} />
    );
    await component.update();

    const cardProps = component.instance().mapTypeContentsToCard(component.state("alerts"),"alerts");
    component.state("alerts").forEach((alert, index) => {
        expect(cardProps[index]).toMatchObject({
            title: alert.date,
            id: alert.key
        });
    })
});

test("Dimensions map correctly to input elements", async () => {
    const component = await shallow(
        <VersionMetadata {...defaultProps} />
    );
    await component.instance().componentWillMount();
    const dimensions = component.instance().mapDimensionsToInputs(component.state("dimensions"));
    let dimensionNames = [];
    dimensions.forEach((dimension, index) => {
        dimensionNames.push({id:dimension.key});
    });

    component.state("dimensions").forEach((dimension, index) => {
        expect(dimensionNames[index]).toMatchObject({
            id: dimension.id
        });
    });
});


test("Changing the dimension title value updates the dimension label in state", async () => {
    const component = await shallow(
        <VersionMetadata {...defaultProps} />
    );
    await component.instance().componentWillMount();
    const mockLabelEvent = {
        target: {
            value: "New label",
            name: "dimension-name",
            id: "time"
        }
    }
    const dimensions = component.instance().mapDimensionsToInputs(component.state("dimensions"));
    expect(component.state("dimensions")[0].label).toBe("");
    component.instance().handleInputChange(mockLabelEvent);
    expect(component.state("dimensions")[0].label).toBe(mockLabelEvent.target.value);
});

// TODO test that the saving responses are passed to the error handler in the correct order