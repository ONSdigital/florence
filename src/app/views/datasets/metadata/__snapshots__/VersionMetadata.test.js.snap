// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dataset version metadata page matches stored snapshot 1`] = `
<div
  className="grid grid--justify-center"
>
  <div
    className="grid__col-xs-10 grid__col-md-6 grid__col-lg-4"
  >
    <div
      className="margin-top--2"
    >
      â—€ 
      <button
        className="btn btn--link"
        onClick={[Function]}
        type="button"
      >
        Back
      </button>
    </div>
    <h1
      className="margin-top--1 margin-bottom--1"
    >
      Metadata
    </h1>
    <p>
      This information is specific to this new data and can be updated each time new data is added.
    </p>
    <div
      className="margin-top--2"
    >
      <div
        className="loader loader--dark"
      />
    </div>
  </div>
</div>
`;
